src/opam.ml

```diff
@@ -18,119 +68,152 @@ module Switch = struct
-        ; Sexp.Atom version
-        ; Sexp.Atom "documentation"
-        ; Sexp.Atom documentation
-        ; Sexp.Atom "synopsys"
-        ; Sexp.Atom synopsis
-        ; Sexp.Atom "path"
-        ; Sexp.Atom path
-        ] ->
-      { name
-      ; version
-      ; path = Path.of_string path
-      ; documentation =
-          ( match documentation with
-          | "" -> None
-          | _ -> Some documentation )
-      ; synopsis =
-          ( match synopsis with
-          | "" -> None
-          | _ -> Some synopsis )
-      }
-    | _ -> assert false
-end
+      Some { path; name; version; items = file_contents }
 
-type t = Cmd.spawn
+  let path t = t.path
 
-let opam_binary = Path.of_string "opam"
+  let name t = t.name
 
-let ocaml_env_binary = Path.of_string "ocaml-env"
+  let version t = t.version
 
-let ocaml_env_setting =
-  Settings.create ~scope:Global ~key:"ocaml.useOcamlEnv"
-    ~of_json:Jsonoo.Decode.bool ~to_json:Jsonoo.Encode.bool
+  let documentation t = Opam_parser.string "doc" t.items
 
-let make () =
-  let spawn =
-    match (Platform.t, Settings.get ocaml_env_setting) with
-    | Win32, Some true ->
-      { Cmd.bin = ocaml_env_binary; args = [ "exec"; "--"; "opam" ] }
-    | _ -> { Cmd.bin = opam_binary; args = [] }
-  in
-  let open Promise.Syntax in
-  let+ spawn = Cmd.check_spawn spawn in
-  match spawn with
-  | Error _ -> None
-  | Ok cmd -> Some cmd
+  let synopsis t = Opam_parser.string "synopsis" t.items
+
+  let depends t =
+    let rec parser = function
+      | [] -> None
+      | OpamParserTypes.Variable (_, s, List (_, l)) :: _
+        when String.equal s "depends" ->
+        let rec aux acc = function
+          | [] -> acc |> List.rev
+          | OpamParserTypes.String (_, v) :: rest -> aux (v :: acc) rest
+          | _ :: rest -> aux acc rest
+        in
+        Some (aux [] l)
+      | _ :: rest -> parser rest
+    in
+    parser t.items
+
+  let has_dependencies t =
+    match depends t with
+    | None
+    | Some [] ->
+      false
+    | _ -> true
+
+  let get_switch_package name ~package_path =
+    let open Promise.Syntax in
+    let* l = Node.Fs.readDir (Path.to_string package_path) in
+    match l with
+    | Error _ -> Promise.return None
+    | Ok l ->
+      Promise.List.find_map
+        (fun fpath ->
+          let basename = Filename.basename fpath in
+          if String.is_prefix basename ~prefix:name then
+            of_path Path.(package_path / fpath)
+          else
+            Promise.return None)
+        l
+
+  let dependencies package =
+    match depends package with
+    | None ->
+      Promise.return
+        (Error "Could not get the root packaged from the switch state")
+    | Some l ->
+      Promise.List.filter_map
+        (fun pkg ->
+          let package_path =
+            (* The package path is never the root, so it's safe to use
+               [value_exn] *)
+            Path.parent (path package) |> fun x -> Option.value_exn x
```

Comment:
`fun x ->` is not needed here.

Reply:
Somehow, `Path.parent (path package) |> Option.value_exn` fails to compile with:

```
This expression has type
  ?here:Base__Source_code_position0.t ->
  ?error:Base.Error.t -> ?message:string -> 'a option -> 'a
but an expression was expected of type Path.t option -> 'b
```

Reply:
Ah yes, I frogot that we use base. I'd suggest you write it as `let x = Option.value_exn (Path.parent (path package))`. `let` bindings are made for this kind of thing.